### PR TITLE
Add scalar_from_constant_expr/scalar_to_constant_expr utilities

### PR DESCRIPTION
**Pull request by steven-johnson**
This was originally part of a PR for a fix for Issue #664, but it
looked fiddly enough that I thought it merited a separate PR on its own.

Basically, Parameter (and thus Argument) use a subset of Expr to hold
the default/min/max values for Generator params; having code to convert
between the Expr-wrapped forms (vs the actual-values we need to create
the metadata) encapsulated in one place makes the rest of the code more
reviewable.

(I’m not entirely happy with this code, as it feels more awkward than
it should be, but it will get the job done.)

### CODE REVIEW

**Comment by abadams** [comment]
This looked painful. If we're inevitably going to need Int64 and Double immediates, maybe we should add them now.

**Comment by steven-johnson** [comment]
Double we can live without. (BTW, don't pull yet, tracking down a subtle bug I missed.)

**Comment by steven-johnson** [comment]
(But yeah, it was painful. That said, the problem here may be that using Expr to hold the def/min/max values wasn't the optimal choice.)

**Comment by steven-johnson** [comment]
Also: adding int64 immediates is going to have an unpleasantly long tail of finding interesting bugs and missed optimizations.

**Comment by steven-johnson** [comment]
FWIW, this is making my spidey-sense tingle a bit, but the alternatives (eg making int64 a first-class node) makes me even more nervous. Is my judgement wrong?

**Comment by abadams** [comment]
The reason I hesitate is that I feel like our inability to represent int64 constants has bitten us before, but I can't recall how.

An alternative that *might* be cleaner is a function call with an implementation in posix math: make_int64(int32_t, int32_t). It'd be easier to unpack, at least.

**Comment by steven-johnson** [comment]
On Tue, Mar 31, 2015 at 8:42 AM, Andrew Adams <notifications@github.com>
wrote:

> An alternative that *might* be cleaner is a function call with an
> implementation in posix math: make_int64(int32_t, int32_t). It'd be easier
> to unpack, at least.
>

Presumably returning an Expr? Not sure I see how it's much cleaner; I guess
rewriting as the Expr equivalent of ((int64_t)hi * (1<<32)) + lo would make
the simplification step more likely to be simplified, and the decomposition
more node-based rather than call-based, but it's not a big needle-mover.

(BTW, simplify doesn't attempt to simplify bitwise shifts or bitwise ors
with constant zero very well. Probably not a big deal.)

**Comment by abadams** [comment]
I mean a runtime function, so it would return an int64_t. The Expr would
just be a call node with two IntImm arguments.

It wouldn't simplify at all, but neither does bit math, as you say.

**Comment by steven-johnson** [comment]
On Tue, Mar 31, 2015 at 9:03 AM, Andrew Adams <notifications@github.com>
wrote:

> I mean a runtime function, so it would return an int64_t. The Expr would
> just be a call node with two IntImm arguments.
>

Calling a runtime function to construct a compile-time constant makes for
cleaner but definitely less-efficient code; we could hope / assume that
LLVM will do the right thing hoisting it, but eww.

Let me take another pass at this today and see if I can simplify / clean it
any further.

If you think adding int64 constants can be done relatively safely and
cleanly, I'll take a stab at it, but it seems like every
IRVisitor/IRMutator will need to be inspected and possibly revised to
handle it.

**Comment by abadams** [comment]
The idea is that the runtime function would be always-inline, so it would
constant-fold away (just as the bit math would). See for example maxval_f64
in posix_math.cpp. I also like that solution because it's where the other
workarounds-for-lack-of-64-bit-immediates live.

I do think that int64 constants can be done relatively safely though. Most
visitors will just correctly ignore them by default. Visitors that must
handle every node generally don't say "using IRVisitor::visit", so with our
compiler flags they'll throw a compile-time error if an implementation is
missing.

**Comment by steven-johnson** [comment]
Reworked to use helper functions for int64 and float64; still not awe-inspiring but considerably less bad. PTAL.

File: src/IROperator.h [hunk 1]
```diff
@@ -1422,6 +1422,166 @@ inline Expr likely(Expr e) {
                                 Internal::vec<Expr>(e), Internal::Call::Intrinsic);
 }
 
+namespace Internal {
+
+namespace {
+
+// Cast e to Type t iff it is not already of that type.
+Expr cast_if(Type t, Expr e) {
```
Comment (abadams) [inline src/IROperator.h]: The existing cast function is already a no-op if the types match.

**Comment by abadams** [comment]
This does indeed feel cleaner to me. I just had that one comment on the use of cast.

File: src/IROperator.h (hunk 1 above)
Comment (steven-johnson) [inline src/IROperator.h]: Oops. Fixed.

**Pull request merged**